crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -1189,6 +1194,110 @@ impl Drop for IndexeddbCryptoStore {
     }
 }
 
+/// Fetch items from an object store in batches, transform each item using
+/// the supplied function, and stuff the transformed items into a single
+/// vector to return.
+async fn fetch_from_object_store_batched<R, F>(
+    object_store: IdbObjectStore<'_>,
+    f: F,
+    batch_size: usize,
+) -> Result<Vec<R>>
+where
+    F: Fn(JsValue) -> Result<R>,
+{
+    enum NextBatch {
+        First,          // In the first batch we don't start from a key
+        Later(JsValue), // In later batches we have a starting key
+    }
+
+    let mut result = Vec::new();
+    let mut next_batch = NextBatch::First;
+    let mut batch_n = 0;
+
+    loop {
+        debug!("Fetching Indexed DB records starting from {}", batch_n * batch_size);
+
+        // See https://github.com/Alorel/rust-indexed-db/issues/31 - we
+        // would like to use `get_all_with_key_and_limit` if it ever exists
+        // but for now we use a cursor and manually limit batch size.
+
+        // Get hold of a cursor for this batch
+        let cursor = match &next_batch {
+            NextBatch::First => object_store.open_cursor(),
+
+            NextBatch::Later(latest_key) => {
+                // We want a cursor starting after the last key we processed
+                let after_latest_key = IdbKeyRange::lower_bound_with_open(latest_key, true);
+
+                // This key should always be valid since we got hold of it
+                // within the same transaction.
+                let after_latest_key = after_latest_key.expect("Key was not valid!");
+
+                object_store.open_cursor_with_range(&after_latest_key)
+            }
+        }?
+        .await?;
+
+        // Fetch batch_size records into result
+        let next_key = fetch_batch(cursor, batch_size, &f, &mut result).await?;
+        if let Some(next_key) = next_key {
```

Comment:
So to avoid the `loop { ... break ... }` pattern, one way would be to:

- roll out the first iteration, where we're opening the cursor
- then use a `while let Some(next_key) = fetch_batch(...).await? { }` loop

It would also avoid the large match by extracting the simple case out of the match (and thus the loop), and the need for the `NextBatch` structure entirely.

Reply:
Great suggestion, but I went with what Rich suggested, using "" as the first key, meaning I don't even need 2 separate cases at all.